### PR TITLE
Refactor Microsoft.CSharp.RuntimeBinder.Semantics.BindingContext

### PR DESCRIPTION
**Pull request by JonHanna**
Remove `BindingContext` fields that always have the same value.

A great many fields are passed around, and sometimes used, but only ever set to the same value.

Remove `OutputContext`, as `m_bUnsafeErrorGiven` is only field remaining and can just be property of `BindingContext`.

Inline the constant values into expressions using them, and remove any code made dead by this.

Since this removes `ExprThisPointer`, `ExprLocal` can now be sealed.

Remove `CheckedContext`: 

Offers no more functionality over the base as the related properties are writeable there.

Rename BindingContextBase.cs to BindingContext.cs

More appropriate, and available after removing `CheckedContext`.

Make `ReportUnsafeErrors` a property, and have it rather than `UnsafeErrorGiven` store state.

Members to properties or auto-properties where appropriate.

Move ctor asserts up to calling factory methods.

Remove no-op assignments of defaults in ctors.

Don't assign to CheckedNormal and CheckedConstant in ctor, as always to false.

Remove static factories and just call ctors.

Tidy up comments.

### CODE REVIEW

File: src/Microsoft.CSharp/src/Microsoft/CSharp/RuntimeBinder/Semantics/BindingContextBase.cs [hunk 1]
```diff
@@ -1,327 +0,0 @@
...
-
-            Debug.Assert(parent.SemanticChecker != null);
-            this.SemanticChecker = parent.SemanticChecker;
-            this.SymbolLoader = SemanticChecker.GetSymbolLoader();
-        }
-
-
-        //The SymbolLoader can be retrieved from m_pSemanticChecker,
-        //but that is a virtual call that is showing up on the profiler. Retrieve
-        //the SymbolLoader once at ruction and return a cached copy.
```
Comment (JonHanna) [inline src/Microsoft.CSharp/src/Microsoft/CSharp/RuntimeBinder/Semantics/BindingContextBase.cs]: I guess someone must have find-replaced away `const` at some point!

**Comment by danmosemsft** [comment]
@dotnet-bot test Innerloop OSX10.12 Release x64 Build and Test please (network)

**Comment by JonHanna** [comment]
Test Innerloop OSX10.12 Release x64 Build and Test

**Comment by danmosemsft** [comment]
@dotnet-bot Test Innerloop OSX10.12 Debug x64 Build and Test (network)
@dotnet-bot  Test Innerloop OSX10.12 Release x64 Build and Test (network)

**Comment by danmosemsft** [comment]
@dotnet-bot Test Innerloop OSX10.12 Debug x64 Build and Test (network)
@dotnet-bot Test Innerloop OSX10.12 Release x64 Build and Test (network)

**Pull request merged**